### ISSUE TITLE
New rule: enforce async/await consistency

### ISSUE THREAD

**Issue by drochgenius** [eslint/eslint#9787]
**Please describe what the rule should do:**

- Enforce that all functions marked as *async* are being awaited when called.
- Enforce that all functons that are being awaited be marked as *async*.
- Prohibit awaiting functions that are not returning promises.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
<!-- put your code examples here -->
function foo1(){  };
async function foo2(){ };

async function bar1(){
    const x = foo1(); // ok
    const y = foo2(); // error
}

async function bar2(){
    const x = await foo1(); // error
    const y = await foo2(); // ok
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Because I think this is core to the language.

**Comment by platinumazure** [comment]
I'm not sure how this rule could deal with functions defined in a different module/script: since ESLint operates on one file at a time, it would be very difficult to know if a function exported from another module/script is in fact async or not.

**Comment by Alexsey** [comment]
And how about
```js
async baz1 () {}

baz2 () {return baz()}

async baz3 () {
  const a = baz2()
}

async baz4 () {
  const a = await baz2()
}
```
`baz3` or `baz4` should be an error?

**Comment by drochgenius** [comment]
Hi @Alexsey ,

I would say `baz3` is a `warning` because the function should not be marked as `async` since it does not contain anything asynchronous inside.

I would say `baz4` is an `error` because `baz2` is not `async`, i.e. not returing a promise. I think it's confusing to await a function that's not asynchronous.

What do you think? Does this make sense to you?

**Comment by Alexsey** [comment]
@drochgenius ok, lets change `baz3` to be like
```js
async baz3 () {
  await Promise.resolve()
  const a = baz2()
}
```
to take a `warning` out of discussion

Now about `baz4` - it is the behavior you would like to reach because `baz2` **would** return a promise and if you want to use it's value you would need to `await` it. And `baz3` on the other hand would assign a promise to `a` and that is something you want to avoid

And that is because promise is a regular value and non-async functions could also return promises and on the other hand you don't need to `await` a promise in 100% of cases

**Comment by drochgenius** [comment]
I understand @Alexsey, and thanks for mentioning this as we wouldn't want to flag an error on valid use cases.

In the light of what you said, the first rule may be just a warning, as it's a potential mistake, but can also be alright.

* _Enforce that all functions marked as async are being awaited when called_

The second rule proposed above are probably more style-related. This is something that I personally try to enforce in my codebases, although there aren't any requirements for that in Javascript.

* _Enforce that all functons that are being awaited be marked as async_.

And the third rule is also style, but it prevents from writing confusing code:

* Prohibit awaiting functions that are not returning promises.

I understand that , at the moment, there is nothing in the language that requires a function that returns a Promise to be marked as `async`, and we don't want either to force people to await on Promises all the time. 

As I'm seeing developers around me adopting `async/await`, they start refactoring complex Promise chains to async await and often they end up with a function not working as expected because they forgot to await on a Promise inside some async function. Would be nice if ESlint could flag these potential errors to the user, but I am not sure what could be the exact form for the rule.

Regards

**Comment by aymericbouzy** [comment]
I'm also interested by having this set of rules.

I guess the example provided by @Alexsey is rather : 

```js
async function baz1 () {}

function baz2 () {return baz1()}

async function baz3 () {
  const a = baz2()
}

async function baz4 () {
  const a = await baz2()
}
```

In any case, I would say in this example that `baz2` should be in error (not awaiting an async function, which by the way makes it incompatible with rule `no-return-await`). Once corrected (awaiting `baz1`, which means `baz2` would now be marked as `async`), `baz3` would be in error, which is the intended behaviour.

**Comment by ljharb** [comment]
Wait, why would `baz2` need to await `baz1`? Similarly, why would it ever be an error to not await a promise in an `async function`? Promises are used all the time without being `await`ed in `async functions`, in particular with `Promise.all`.

**Comment by aymericbouzy** [comment]
`Promise.all` and `Promise.race` are situations where you obviously not await promises.

`baz2` doesn't "need" to await `baz1`, but allowing this type of code leads to difficult to maintain code (code that doesn't look asynchronous, but has asynchrony in it hidden away).

My really pain point is the following : I have a large codebase, and when my code runs, I encounter bugs that are difficult to reproduce, unpredictable results ... simply because I forgot to await a promise. Here is a simple example where not awaiting a promise leads to bad results:

```js
let foo

async function bar() {
	// does some stuff with foo
}

export default async function(someObject = {}) {
	foo = someObject
	bar()
	foo = null
}
```
The issue I have is that the error does not tell me where I accidentally forgot the `await` keyword. I have to scan my files and find the place where I made the blunder. If ESlint could tell me "Wait, this is weird, you are not awaiting on this asynchronous function", it would ease my life a lot.

**Comment by Alexsey** [comment]
@aymericbouzy I have set up my Intellij Webstorm to highlight all usages of async functions without await. So it is not an error but when you are looking at the code you can explicitly see all that places. Probably you can try to set up your IDE/text editor somehow similar. But it can't be an error

**Comment by aymericbouzy** [comment]
Not sure how to do it in VS Code ... still looking around, found this in the meantime : https://github.com/palantir/tslint/pull/1632

**Comment by drochgenius** [comment]
Hi @aymericbouzy ,  

This is exactly the same issue I am facing all the time as my team starts to use async/await instead of Promises.
This said, I perfectly understand that this cannot be an error as there are valid use cases for not awaiting on promises. 
Also, as @platinumazure  mentioned above, ESLint parses one file at a time, it would be difficult to identify this kind of pattern across module imports. So maybe ESLint is not the right place to implement this kind of rule.
If you find a way to setup VSCode to flag this potential mistake, please leave a note here.
I'll do the same on my side. We could maybe write a VSCode extension for this.

Regards

**Comment by aymericbouzy** [comment]
@drochgenius I think we have to turn to extensions of javascript that have types, such as TypeScript or Flow 😕

**Comment by ljharb** [comment]
This is what a type system and/or tests are for; this just isn’t a lintable problem.

**Comment by aymericbouzy** [comment]
@ljharb you mention tests : how would you write a test that catches this type of error? I'm curious, this could solve our problem!

**Comment by ljharb** [comment]
Your test framework should be able to handle async verification; you’d await the result of the promise, and then assert on the state of the world.

It wouldn’t catch race conditions reliably, but that’s a separate problem.

**Comment by drochgenius** [comment]
Fine by me, thanks.
